core/store/src/trie/node_storage.rs

```diff
@@ -1,24 +1,69 @@
-use super::mem::updating::{
-    GenericNodeOrIndex, GenericTrieUpdate, GenericUpdatedTrieNode, UpdatedTrieStorageNodeWithSize,
-};
-use super::TrieRefcountDeltaMap;
-use crate::trie::{
-    Children, RawTrieNode, RawTrieNodeWithSize, StorageHandle, StorageValueHandle, ValueHandle,
-};
-use crate::{StorageError, Trie, TrieChanges};
 use borsh::BorshSerialize;
+use near_primitives::errors::StorageError;
 use near_primitives::hash::{hash, CryptoHash};
 use near_primitives::state::ValueRef;
 
+use super::ops::interface::{
+    GenericNodeOrIndex, GenericTrieUpdate, GenericTrieValue, GenericUpdatedNodeId,
+    GenericUpdatedTrieNode, GenericUpdatedTrieNodeWithSize,
+};
+use super::{
+    Children, RawTrieNode, RawTrieNodeWithSize, StorageHandle, StorageValueHandle, Trie,
+    TrieChanges, TrieRefcountDeltaMap, ValueHandle,
+};
+
+const INVALID_STORAGE_HANDLE: &str = "invalid storage handle";
+
+pub(crate) type TrieStorageNodePtr = CryptoHash;
+
+pub(crate) type UpdatedTrieStorageNode = GenericUpdatedTrieNode<TrieStorageNodePtr, ValueHandle>;
+
+impl UpdatedTrieStorageNode {
+    fn new_branch(children: Children, value: Option<ValueRef>) -> Self {
+        let children =
+            Box::new(children.0.map(|child| child.map(|id| GenericNodeOrIndex::Old(id))));
+        let value = value.map(ValueHandle::HashAndSize);
+        Self::Branch { children, value }
+    }
+
+    /// Conversion from the node just read from trie storage.
+    pub fn from_raw_trie_node(node: RawTrieNode) -> Self {
+        match node {
+            RawTrieNode::Leaf(extension, value) => Self::Leaf {
+                extension: extension.to_vec().into_boxed_slice(),
+                value: ValueHandle::HashAndSize(value),
+            },
+            RawTrieNode::BranchNoValue(children) => Self::new_branch(children, None),
+            RawTrieNode::BranchWithValue(value, children) => {
+                Self::new_branch(children, Some(value))
+            }
+            RawTrieNode::Extension(extension, child) => Self::Extension {
+                extension: extension.to_vec().into_boxed_slice(),
+                child: GenericNodeOrIndex::Old(child),
+            },
+        }
+    }
+}
+
+pub(crate) type UpdatedTrieStorageNodeWithSize =
+    GenericUpdatedTrieNodeWithSize<TrieStorageNodePtr, ValueHandle>;
+
+impl UpdatedTrieStorageNodeWithSize {
+    pub fn from_raw_trie_node_with_size(node: RawTrieNodeWithSize) -> Self {
+        Self {
+            node: UpdatedTrieStorageNode::from_raw_trie_node(node.node),
+            memory_usage: node.memory_usage,
+        }
+    }
+}
+
 pub(crate) struct NodesStorage<'a> {
```

Comment:
Would it be possible to add a short comment to explain the usage and/or purpose of this struct?

Reply:
Will do in a follow up PR